test/Test/Pos/Block/Logic/VarSpec.hs

```diff
@@ -168,29 +184,102 @@ blockEventSuccessSpec = do
         "a sequence of interleaved block applications and rollbacks " <>
         "results in the original state of the blockchain"
 
-blockEventSuccessProp :: BlockProperty ()
-blockEventSuccessProp = do
+{- | This generator is carefully designed to cover multitude of success
+   scenarios. Apply/rollback are interleaved in various ways, shown by diagrams below:
+
+   0 -----a----> 2
+   |             |        Synchronous apply/rollback
+   0 <----r----- 2
+
+   0 -a-> 1 -a-> 2
+   |             |        Multiple apply per rollback
+   0 <----r----- 2
+
+   0 -----a----> 2
+   |             |        Multiple rollback per apply
+   0 <-r- 1 <-r- 2
+
+   0 -a-> 3 -a-> 6 -a-> 9
+   |                    |        Desynchronous apply/rollback
+   0 <--r--- 4 <---r--- 9
+
+   Furthermore, it allows nested forks (forks of forks), generates both unique
+   and repeated forks, respects the 'blkSecurityParam', and can be used with
+   'enrichWithSnapshotChecking'. (I would draw diagrams for these features as
+   well, but they're barely readable in the ASCII format). Just trust me that
+   this generator gives diverse block event sequences -- I spent an entire night
+   and a few sheets of paper trying to figure out how to write it.
+-}
+
+genSuccessWithForks :: forall g m. (RandomGen g, Monad m) => BlockEventGenT g m ()
+genSuccessWithForks = do
+      emitBlockApply BlockApplySuccess $ pathSequence mempty ["0"]
+      generateFork "0" []
+      emitBlockApply BlockApplySuccess $ pathSequence "0" ["1", "2"]
+      generateFork ("0" <> "1" <> "2") []
+  where
+    generateFork ::
```

Comment:
I hope this logic is correct. it's so gribnaya that I can't force myself to delve into it.